apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/BodyCaptureImpl.java

```diff
@@ -0,0 +1,136 @@
+package co.elastic.apm.agent.impl.context;
+
+import co.elastic.apm.agent.objectpool.Resetter;
+import co.elastic.apm.agent.objectpool.impl.QueueBasedObjectPool;
+import co.elastic.apm.agent.tracer.configuration.WebConfiguration;
+import co.elastic.apm.agent.tracer.metadata.BodyCapture;
+import co.elastic.apm.agent.tracer.pooling.Allocator;
+import co.elastic.apm.agent.tracer.pooling.ObjectPool;
+import co.elastic.apm.agent.tracer.pooling.Recyclable;
+import org.jctools.queues.atomic.MpmcAtomicArrayQueue;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+
+public class BodyCaptureImpl implements BodyCapture, Recyclable {
+    private static final ObjectPool<ByteBuffer> BYTE_BUFFER_POOL = QueueBasedObjectPool.of(new MpmcAtomicArrayQueue<ByteBuffer>(128), false,
+        new Allocator<ByteBuffer>() {
+            @Override
+            public ByteBuffer createInstance() {
+                return ByteBuffer.allocate(WebConfiguration.MAX_BODY_CAPTURE_BYTES);
+            }
+        },
+        new Resetter<ByteBuffer>() {
+            @Override
+            public void recycle(ByteBuffer object) {
+                object.clear();
+            }
+        });
+
+    private enum CaptureState {
+        NOT_ELIGIBLE,
+        ELIGIBLE,
+        STARTED
+    }
+
+    private CaptureState state;
```

Comment:
Given the description of changing state in one place but capturing in another, I think the `state` access and updates need to made thread-safe?